server/apps/microservices/src/processors/asset-uploaded.processor.ts

```diff
@@ -0,0 +1,67 @@
+import { InjectQueue, OnQueueActive, OnQueueCompleted, OnQueueWaiting, Process, Processor } from '@nestjs/bull';
+import { Job, Queue } from 'bull';
+import { AssetEntity, AssetType } from '@app/database/entities/asset.entity';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { randomUUID } from 'crypto';
+
+@Processor('asset-uploaded-queue')
+export class AssetUploadedProcessor {
+  constructor(
+    @InjectQueue('thumbnail-generator-queue')
+    private thumbnailGeneratorQueue: Queue,
+
+    @InjectQueue('metadata-extraction-queue')
+    private metadataExtractionQueue: Queue,
+
+    @InjectQueue('video-conversion-queue')
+    private videoConversionQueue: Queue,
+
+    @InjectRepository(AssetEntity)
+    private assetRepository: Repository<AssetEntity>,
+  ) {}
+
+  /**
+   * Post processing uploaded asset to perform the following function if missing
+   * 1. Generate JPEG Thumbnail
+   * 2. Generate Webp Thumbnail <-> if JPEG thumbnail exist
+   * 3. EXIF extractor
+   * 4. Reverse Geocoding
+   *
+   * @param job asset-uploaded
+   */
+  @Process('asset-uploaded')
+  async processUploadedVideo(job: Job) {
+    const {
+      asset,
+      fileName,
+      fileSize,
+      hasThumbnail,
+    }: { asset: AssetEntity; fileName: string; fileSize: number; hasThumbnail: boolean } = job.data;
+
+    if (hasThumbnail) {
+      // The jobs below depends on the existence of jpeg thumbnail
+      await this.thumbnailGeneratorQueue.add('generate-webp-thumbnail', { asset }, { jobId: randomUUID() });
+      await this.metadataExtractionQueue.add('tag-image', { asset }, { jobId: randomUUID() });
```

Comment:
tag-image and detect-object should happen regardless of whether the thumbnail exists

Reply:
The function that decodes the image into tensors of TensorFlow expects these formats `BMP, GIF, JPEG, and PNG`. So if you upload a raw formats (HEIC, HEIF, or DNG), this function won't be able to run

Reply:
Yes, but it is also the trade-off between CPU (or GPU) usage. Below is a good read/discussion about this topic

https://stackoverflow.com/questions/45149023/tensorflow-for-image-recognition-size-of-images

Reply:
That's talking specifically about training the model, right? Presumably you're using a pre-trained model and just pushing images into it for analysis and tagging, I would imagine for that task using the full resolution image is preferable, especially if that's what the data was trained on.

Reply:
Correct, the full resolution would take more CPU to run through all the pixels, so it is the trade-off between accuracy and CPU usage. I guess we can also throttle this queue with the concurrency level as well. I guess we can have this feature in the future as a setting in the admin panel as well since other people might have their own model that takes a specific resolution. So the ML pipeline can generate the "thumbnail" image with the specified resolution from the user.

Reply:
Yea for sure, would be interested to see how much more CPU time it takes to run on the full resolution image, with my understanding of how the models tend to work I wouldn't expect it to actually be that much more intensive for analysis, just for training. Can take a look at that later though, was just interesting to see how this works.